### PR TITLE
Add decorator functionality

### PR DESCRIPTION
**Pull request by roykillany**
## Background
The particular use case of presenting/decorating the fetched records is one that is not currently accounted for. I thought it would make good sense to be able to supply some decorator that the `Paginator` could use to change the returned data.

## Changes
- [ ] update readme
- [ ] update change log
- [x] add `decorator` kwarg
- [x] calls `decorator` on each `item` (https://github.com/xing/rails_cursor_pagination/pull/27/files#diff-c201f4d60dd83e98535d880e8d9b9ad479792383bb84e0cddd4a9122514e11b5R159)
- [ ] add contributors

### CODE REVIEW

File: spec/rails_cursor_pagination/paginator_spec.rb [hunk 1]
```diff
@@ -274,6 +281,28 @@
...
+        end
+      end
+
+      context 'when passing a valid lambda as `&record_decorator`' do
+        let(:id_incrementer) { ->(item) { { next_id: item.id + 1 } } }
+        let(:params) { super().merge(record_decorator: id_incrementer) }
+        let(:expected_decorated_response) do
+          expected_posts.map do |post|
+            id_incrementer.call(post)
+          end
```
Comment (svendittmer) [inline spec/rails_cursor_pagination/paginator_spec.rb:299]: 💅 
```suggestion
          expected_posts.map(&id_incrementer)
```

File: lib/rails_cursor_pagination/paginator.rb [hunk 2]
```diff
@@ -99,10 +105,12 @@ def fetch(with_total: false)
     #   Optional, cannot be combined with `after`
     # @param order [Symbol]
     #   Optional, must be :asc or :desc
+    # @param record_decorator [Object, proc { |obj| obj.itself }]
+    #   Optional, must respond to `.call`
```
Comment (svendittmer) [inline lib/rails_cursor_pagination/paginator.rb:109]: Passing a `record_decorator` is optional in `#initialize`, but `#ensure_valid_params!` actually requires it.

**Review by nicolas-fricke — COMMENTED** [review]
Thanks for the contribution @roykillany! As already discussed before, I think this is a super valuable addition to the gem.

We also talked about what the right place to pass this decorator would be, and while it might make more sense to actually pass it to the `#fetch` method, it would also require some bigger refactoring regarding the memoization. I suppose that is something that can be done when we tackle https://github.com/xing/rails_cursor_pagination/issues/1 during which we'll also further split up this class. That will likely have to be a breaking change.

What I think is still missing from this PR before we can merge it is a section in the [README.md](https://github.com/xing/rails_cursor_pagination/blob/master/README.md#usage) in the _**Usage**_ section detailing this new option and explaining how and why to use it. E.g. something like:
```markdown
### Decorating the response

By default, the hash returned by the `#fetch` method will just contain the plain `ActiveRecord::Base` model instances that were fetched from the database.
Using a `#select` statement on the query, we can decide what fields should be fetched and thereby somewhat reduce the data returned.

But often times we might want to be able to also expose data that e.g. is calculated by a method or transform a given datum before returning it.
That is where the `record_decorator` can come in handy.
It allows us to pass a proc / lambda or any other callable object which will then be invoked for each record that is retrieved from the database.
Whatever this proc returns will then be returned as `data` on the `page` objects.

For an example, assume that our `Post` model has a method `#urn` that returns a unique string identifying each post based on its type and ID.
And imagine we want to expose this information along with the `author` from the DB but call the field `name`.
To achieve that, we can define a proc that will do just that:
```ruby
decorator = ->(record) { { urn: record.urn, name: record.author } }

RailsCursorPagination::Paginator
  .new(posts, record_decorator: decorator)
  .fetch
\```

This would then expose return something similar to this response:
```ruby
{
  page_info: {
    has_previous_page: false,
    has_next_page: true,
    start_cursor: "MQ==",
    end_cursor: "MTA="
  },
  page: [
    { cursor: "MQ==", data: { urn: "post:1", author: "Sam" } },
    { cursor: "Mg==", data: { urn: "post:2", author: "Jace" } },
    ...
  ]
}
\```
```

And then a quick update to the [`CHANGELOG.md` -> Unreleased](https://github.com/xing/rails_cursor_pagination/blob/master/CHANGELOG.md#unreleased) section detailing the change, and from my side, we should be good to merge it :)

**Comment by roykillany** [comment]
ill reopen this when i have time to work on it

**Pull request closed without merging**